fastdata/core.py

```diff
@@ -44,23 +102,123 @@ def generate(self,
+            commit_every (Union[int, float], optional): The number of minutes between each commit. Defaults to 5.
+            private (bool, optional): Whether the repository is private. Defaults to False.
+            token (Optional[str], optional): The token to use to commit to the repo. Defaults to the token saved on the machine.
+            delete_files_after (bool, optional): Whether to delete files after processing. Defaults to True.
+
+        Returns:
+            tuple[str, list[dict]]: A tuple with the generated repo_id and the list of generated data dictionaries.
+        """
+        dataset_dir = Path(repo_id.replace("/", "_"))
+        dataset_dir.mkdir(parents=True, exist_ok=True)
+        data_dir = dataset_dir / "data"
+        data_dir.mkdir(exist_ok=True)
+
+        try:
+            scheduler = CommitScheduler(
+                repo_id=repo_id,
+                repo_type="dataset",
+                folder_path=dataset_dir,
+                every=commit_every,
+                private=private,
+                token=token,
+            )
+            DatasetCard(
+                DATASET_CARD_TEMPLATE.format(
+                    title=repo_id,
+                    model_id=self.cli.model,
+                    system_prompt=sp,
+                    prompt_template=prompt_template,
+                    sample_input=inputs[:2],
+                )
+            ).save(dataset_dir / "README.md")
+
+            results = []
+            total_inputs = len(inputs)
+
+            with tqdm(total=total_inputs) as pbar:
+                with concurrent.futures.ThreadPoolExecutor(
+                    max_workers=max_workers
+                ) as executor:
+                    futures = [
+                        executor.submit(
+                            self._process_input,
+                            prompt_template,
+                            schema,
+                            temp,
+                            sp,
+                            input_data,
+                        )
+                        for input_data in inputs
+                    ]
+                    completed = 0
+
+                    for completed_future in concurrent.futures.as_completed(futures):
+                        result = completed_future.result()
+                        if result is not None:
+                            results.append(result)
+                        completed += 1
+                        pbar.update(1)
+
+                        if completed % save_interval == 0 or completed == total_inputs:
+                            if results:
+                                with scheduler.lock:
+                                    self._save_results(
+                                        results, data_dir / f"train-{uuid4()}.jsonl"
+                                    )
+                                results.clear()
+                    scheduler._push_to_hub()
+                    scheduler.stop()
```

Comment:
The scheduler is meant to upload files at regular interval without taking care of it once initialized. The implementation here works but is not better than simply using `upload_folder` directly. What I would suggest to do is to save results and let the scheduler upload them every 5 minutes as planned above with `every=commit_every,`. 

I'd suggest a logic like this:

```suggestion
                    current_file = data_dir / f"train-{uuid4()}.jsonl"
                    for completed_future in concurrent.futures.as_completed(futures):
                        result = completed_future.result()
                        if result is not None:
                            results.append(result)
                            with scheduler.lock:
                                self._save_results(results, current_file)

                        pbar.update(1)

                        if len(results) >= max_items_per_file:
                            current_file = data_dir / f"train-{uuid4()}.jsonl"
                            results.clear()
```

In particular:
- each time a future is completed, append the result to the jsonl file. Here I'm using `_save_results` which is suboptimal as we could only append the data to the file but I guess the I/O optimization is negligible for now
- (optional) define a `max_items_per_file` to set a limit to the number of items per file (avoids having a single gigantic file)
- no need for `save_interval` parameter and `completed` counter
- no need for calling `scheduler._push_to_hub` and `scheduler.stop` => results are saved to disk immediately once a future is completed and uploaded to the Hub by chunks once every 5 minutes. 
- since you use `scheduler.lock`, there is no risk of concurrency between the serialization to disk and the upload (i.e. cannot upload a corrupted file)

Let me know if my explanation was clear enough. This is not a feature used so often (even though it's fully maintained!) so happy to get some feedback.

Reply:
Thanks, @Wauplin! I incorporated your suggestions in the latest PR. I didn't apply them directly here because this file is auto-generated from modifications made to `nbs/00_code.ipynb`